File: Code/Python/queue.py
"""
Problem Statement:
Add implementation of queue in Python.

Queue: A queue is a linear data-structure where elements can be inserted and deleted from both the ends.

A queue has the following properties.

• It is opened at both the ends.

• The end at which elements are addded in a queue is called the rear/back/tail.

• The end from which elements are removed from the queue is called head/front.

• Queue follows FIFO/LILO methodology to insert/remove elements.

• Unlike lists/arrays, queues don't allow for random access to the elements.

Main Operations in Queue:
• Enqueue: Used to add items to the queue. If the queue is full, it is a condition of the Queue Overflow. Time complexity: O(1).

• Dequeue: Used to remove an element from the queue.If the queue is empty, it is a condition of the Queue Underflow. Time complexity: O(1).

• Front/Peek: Displays an element from the front end. Time complexity: O(1).

• Rear: Displays an element from the rear end. Time complexity: O(1).

• isEmpty: Returns True if the Queue is empty, otherwise, False.
"""


class queue:
    def __init__(self):
        self.__queue = []
        self.__size = 0

    def enqueue(self, data):
        """ adds items to the queue """
        self.__queue.append(data)
        self.__size += 1

    def dequeue(self):
        """ remove items from the queue """
        if not self.isEmpty():
            element = (self.__queue[0])
            self.__queue.remove(element)
            self.__size -= 1
            print(element, "has been removed")
            return

    def front(self):
        """ displays an item from the front end """
        print("Front element is:", self.__queue[0])
        return

    def rear(self):
        """ displays an item from the rear end """
        print("Rear element is :", self.__queue[-1])
        return

    def isEmpty(self):
        """ returns true if the queue is empty """
        return self.__size == 0

    def display_queue(self):
        """ displays the queue """
        if self.isEmpty():
            print("The queue is empty")
            return
        for i in self.__queue:
            print(i, "--->", end=' ')
        print(None)


if __name__ == "__main__":
    q = queue()
    print("Enter values to push in the queue: ")
    value=[ele for ele in input().split()]
    for i in range(len(value)):
        q.enqueue(value[i])
    q.display_queue()
    q.front()
    q.dequeue()
    q.front()
    q.rear()
    print(q.isEmpty())

"""Time-Complexity of the program:
enqueue(): O(1)
dequeue(): O(1)
front(): O(1)
rear(): O(1)
isEmpty(): O(1)
display_queue(): O(N)
"""

"""Test-Cases:

>>>
Input:
Enter values to push in the queue:
5 6 7

Output:
5 ---> 6 ---> 7 ---> None
Front element is: 5
5 has been removed
Front element is: 6
Rear element is : 7
False

>>>
Input:
Enter values to push in the queue:
Anamika B.Tech Computer-Science

Output:
Anamika ---> B.Tech ---> Computer-Science ---> None
Front element is: Anamika
Anamika has been removed
Front element is: B.Tech
Rear element is : Computer-Science
False

>>>
Input:
Enter values to push in the queue:
1 Apple Bangkok 32 64

Output:
1 ---> Apple ---> Bangkok ---> 32 ---> 64 ---> None
Front element is: 1
1 has been removed
Front element is: Apple
Rear element is : 64
False
"""

File: Queue/queue.java
/*
 *                    Implementation of Queue
 *         
 * Queue is a linear structure which follows a particular order in 
 * which the operations are performed. 
 * The order is First In First Out (FIFO).
 */
import java.util.*;
class Queue {
    int front, rear, size;
    int capacity;
    int array[];

    public Queue(int capacity)
    {
        this.capacity = capacity;
        front = this.size = 0;
        rear = capacity - 1;
        array = new int[this.capacity];
    }

    boolean isFull(Queue queue)
    {
        return (queue.size == queue.capacity);
    }

    boolean isEmpty(Queue queue)
    {
        return (queue.size == 0);
    }

    void enqueue(int item)
    {
        if (isFull(this))
            return;
        this.rear = (this.rear + 1)        % this.capacity;
        this.array[this.rear] = item;
        this.size = this.size + 1;
        System.out.println(item
            + " enqueued to queue");
    }

    int dequeue()
    {
        if (isEmpty(this))
            return Integer.MIN_VALUE;

        int item = this.array[this.front];
        this.front = (this.front + 1)        % this.capacity;
        this.size = this.size - 1;
        return item;
    }

    int front()
    {
        if (isEmpty(this))
            return Integer.MIN_VALUE;

        return this.array[this.front];
    }

    int rear()
    {
        if (isEmpty(this))
            return Integer.MIN_VALUE;

        return this.array[this.rear];
    }

    public static void main(String[] args)
    {
        Scanner in=new Scanner(System.in);
        Queue queue = new Queue(1000);
        System.out.println("No. of elements to be added in Queue :");
        int n=in.nextInt();
        for(int i=0;i<n;i++)
        {
            int input=in.nextInt();
            queue.enqueue(input);

        }

        System.out.println("Front item is "
            + queue.front());
        System.out.println(queue.dequeue()
            + " dequeued from queue\n");
        System.out.println("Front item is "
            + queue.front());

        System.out.println("Rear item is "
            + queue.rear());

            in.close();
    }
}

/*
 * Test Case :
 * Input : No. of elements to be added in Queue :
 * 5 
 * 1 2 3 4 5
 * Output :
 * 1 enqueued to queue
   2
   2 enqueued to queue
   3
   3 enqueued to queue
   4
   4 enqueued to queue
   5
   5 enqueued to queue
   
   Front item is 1
   1 dequeued from queue

   Front item is 2
   Rear item is 5
   
 * Time  Complexity is
 * Insertion(Enqueue) O(1) 
 * Deletion(Dequeue)  O(1) 
 * Space Complexity is O(n) , n is the size of the Queue.
 */